Create new component TreeSelect and migrate category field

nathanss opened the pull request:
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Some time ago the Tree control was written and we decided to migrate the category field to it. 

Previously, the Select control was used for the input, with a callback to render the children as a popover. I had a lot of problems trying to make Select work with TreeControl, so I decided to create a new component to wrap this complexity and also stopped using SelectControl.

Closes #37043.
Closes #35969. 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Please include detailed instructions on how these changes can be tested, make sure to review and follow the guide for writing high-quality testing instructions below. -->

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

1. Run `pnpm install` and `pnpm build`
2. Open `/wp-admin/admin.php?page=wc-admin&path=%2Fadd-product-old`
3. Go to the Category field.
4. Test general usage, both with keyboard and mouse.
5. Compare to the `trunk` version. Check if you see any regressions.

<!-- End testing instructions -->

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [ ] Have you included testing instructions?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

nathanss committed:
Add initial implementation of popover using TreeControl

nathanss committed:
Expose TreeItem interface

nathanss committed:
Add onSelect and onRemove handlers and provisory width

nathanss committed:
Create `shouldNotRecursivelySelect` prop to adjust the behavior needed for category field

nathanss committed:
Add create new functionality to Tree

nathanss committed:
Hide create new when exact value is typed

nathanss committed:
Push to Fernando

nathanss committed:
Start experiment of including downshift's props in Tree

nathanss committed:
Try to make TreeControl work with downshift's props

nathanss committed:
Revert changes in TreeControl and related hooks

nathanss committed:
Create new component to replace Select

nathanss committed:
Add more functionality to SelectTree

nathanss committed:
Restore CSS

nathanss committed:
Prevent createvalue from being added to the DOM

nathanss committed:
Add improvements to select-tree component

Use Dropdown instead of Popover
Improve handling of when the dropdown should be open

nathanss committed:
Rename fields from Tree to have better compatibility with Category Field

nathanss committed:
Make filtering items works

nathanss committed:
Close popover on tab

nathanss committed:
Rename old field value

nathanss committed:
Changelog

nathanss committed:
Fix bug in combo-box when ref passed is not a callback
Improve some type definitions
Remove old component
Add loading state

nathanss committed:
Refactor

nathanss committed:
Remove now unneeded preventClose ref

nathanss committed:
Remove unneeded CSS

nathanss committed:
Prevent scrolling when focus is on Create button

nathanss committed:
Improve query to focus on first element

nathanss committed:
Add __experimental to SelectTree

nathanss committed:
Fix lint

nathanss committed:
Add props documentation

louwie17 commented on packages/js/components/src/experimental-tree-control/hooks/use-tree.ts:
```diff
@@ -19,6 +19,8 @@ export function useTree( {
 	onSelect,
 	onRemove,
+	shouldNotRecursivelySelect,
+	createValue,
```
`createValue` is not used, or are you doing this explicitly to not pass it down into `treeProps`?

nathanss replied:
Explicitly to not pass it to treeProps... That's a clear case where the mass use of the spread operator makes the code more complicated. I would love to ban them from the codebase 😂

louwie17 commented on packages/js/components/src/experimental-tree-control/select-tree.tsx:
```diff
@@ -0,0 +1,193 @@
+							ref: comboBoxRef,
+						} }
+						inputProps={ {
+							className:
```
This will also need some `aria` labels for combobox in order to support screen readers. Like: `aria-autocomplete="list" aria-controls="downshift-0-menu" aria-labelledby="downshift-0-label" autocomplete="off"`
Might be worth what `downshift` is doing with the experimental select control (that's where the above comes from).

louwie17 replied:
The `aria-controls` will have to match an `id` on the tree control wrapper.
See here: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/combobox_role
for more info.

nathanss replied:
Thanks for pointing this out! I was doing AB testing yesterday at the end of the day and I noticed that the screen reader stopped giving some additional information. I will look into it.

louwie17 commented on packages/js/components/src/experimental-tree-control/tree.tsx:
```diff
@@ -20,24 +23,60 @@ export const Tree = forwardRef( function ForwardedTree(
+					/>
+				) ) }
+			</ol>
+			{ props.allowCreate && ! createValueExists && (
```
Given this is not part of the `TreeItem` data it doesn't allow a user to select it using the keyboard, would this still be possible outside of the `TreeItem` component?

nathanss replied:
I will take a look. Maybe I could use a "fixed Tree Item" for this functionality

louwie17 commented on packages/js/product-editor/src/components/details-categories-field/create-category-modal.tsx:
```diff
@@ -91,7 +92,7 @@ export const CreateCategoryModal: React.FC< CreateCategoryModalProps > = ( {
 					onChange={ setCategoryName }
 				/>
-				<SelectControl< Pick< ProductCategory, 'id' | 'name' > >
+				<SelectControl< ProductCategoryLinkedList >
```
This should probably also be replaced by the new TreeSelect, although that can be done in a follow up.

nathanss replied:
Yes. I plan on creating another issue when this one is merged.

louwie17 commented on packages/js/product-editor/src/components/details-categories-field/use-category-search.ts:
```diff
@@ -272,9 +278,33 @@ export const useCategorySearch = () => {
+		getFilteredItemsForSelectTree,
 		categoriesSelectList: categoriesAndNewItem[ 0 ],
 		categories: categoriesAndNewItem[ 1 ],
 		isSearching,
```
A lot of this logic can be removed once we also replace the `TreeSelect` in the add new modal, as the `TreeControl` will handle a lot of this for us.

louwie17 commented on packages/js/components/src/experimental-tree-control/hooks/use-linked-tree.ts:
```diff
@@ -9,7 +9,7 @@ import { useMemo } from 'react';
 
 type MemoItems = {
-	[ value: Item[ 'value' ] ]: LinkedTree;
+	[ value: Item[ 'id' ] ]: LinkedTree;
```
Is there a particular reason you changed to `id` instead of `value`?
Not a huge issue, but curious if there was a specific reason for this.

nathanss replied:
I thought that since this is the first use case for TreeControl, and the equivalent data is called `id` on the `useCategorySearch` hook, it made sense to name it equally.

louwie17 reviewed (changes requested):
Nice work @nathanss, this is looking quite good, thanks for willing to tackle this, as it is a big overhaul!
I did leave several comments within the code, I also noticed a few small issues, but nothing major:
- Selecting the `Create new` item is not possible with the keyboard
- When searching for a value that is nested the parent doesn't automatically open, I thought this was something the original TreeControl did by default.
- It appears that all the TreeControl storybooks are broken now.

louwie17 commented:
Oooh also @nathanss, this branch seems quite outdated with `trunk`, it would probably be worth rebasing it :)

nathanss committed:
Add more aria props

nathanss committed:
Remove wrongfully passed props to elements
Enable scrolling between button and tree with up and down arrows

nathanss committed:
Fix storybook

nathanss committed:
Expand parent when children is being searched

nathanss commented:
@louwie17 thanks for the thorough review! I think I have addressed all your feedback. I'm not so happy with the solution to the create button using the arrow keys, but I can't think of an alternative.

Do you think it's possible to have only one <li> from inside a <ol> show in a fixed position outside the scroll, as it currently looks right now? This would make a lot of things easier. But I don't think that's possible.

louwie17 commented on packages/js/components/src/experimental-tree-control/select-tree.tsx:
```diff
@@ -0,0 +1,218 @@
+							ref: comboBoxRef,
+							role: 'combobox',
+							'aria-expanded': isOpen,
+							'aria-haspopup': 'listbox',
```
The value for this should probably be `tree`

louwie17 commented on packages/js/components/src/experimental-tree-control/tree.tsx:
```diff
@@ -20,24 +23,84 @@ export const Tree = forwardRef( function ForwardedTree(
+								(
+									document.querySelector(
+										'.experimental-woocommerce-tree__button'
+									) as HTMLButtonElement
```
Instead of using `document` here could you make use of `ref.current.parentElement`?
As this may focus other create new buttons if another tree happens to be rendered on the same page.

While this is an unlikely scenario, it probably would still be worth addressing.

louwie17 replied:
cc @mdperez86 curious what you think about this approach with highlighting the create new item, and if we are better of modifying the Tree data structure to support the addition of a create new item within the `items` array, so it can be more native to the `Tree` component.

I am ok with this for now, but it does feel like a bit of hacky solution.

louwie17 reviewed (commented):
Thanks for the updates and fixing the storybooks 🎉 💯 

Thanks for addressing the screen reader stuff and adding some `aria` labels, I think it does seem to announce some more stuff now, but still not the results or whether the popover actually opened or not.
I am not a `aria` expert though, so given it does announce the selection and such I am fine on merging this part as is, but cc'ing @joshuatf as I know he is a bit more well versed in the aria stuff.
We can always tackle more of this in follow ups.

I also cc'd @mdperez86 on the create new selection approach, I would still favour it being integrated as part of the tree control somehow. Seeing if Maikel may have some suggestions as he was the Tree component creator. 
But your current approach does fix the selection and seems to work relatively well ( I did leave one suggestion on `querySelector` your using).

For the rest, I only have one more suggestion as a blocker, aside from the additional smaller comments within the code.
- Currently when not searching anything and expanding a field and selecting a child, the parent automatically closes. This feels kinda weird from a user perspective, is it possible to keep the parent open upon selection? Maybe this should be an optional prop set on the `SelectTree` component.

nathanss committed:
Change haspopup aria label

nathanss committed:
Use ref instead of querying all the document to focus on button

nathanss commented:
@louwie17 I agree with the behavior of keeping the parent open when there's a child selected. Can we (I) create a separate issue for that after we merge this? Because it seems like I'll have to touch at least two hooks for that `useExpander` and `useSelection`. Currently they seem to be separated.
